python/lsst/ci/models.py

```diff
@@ -0,0 +1,138 @@
+from dataclasses import dataclass
+from typing import List, Set, Optional, OrderedDict as ODict
+from collections import OrderedDict
+
+from lsst.ci import tsort
+
+
+class ProductIndex(OrderedDict):
+    """
+    Container class for a product index with index-related operations.
+    """
+
+    def __init__(self, *args, **kwargs):
+        """Create a new product index object.
+
+        Parameters
+        ----------
+        product_index
+        """
+        super().__init__()
+        self.toposorted = False
+        self.sorted_groups = []
+
+    def toposort(self) -> ODict[str, "Product"]:
+        """Topologically sort the product index and return it.
+
+        This mutates this object.
+
+        Returns
+        -------
+        OrderedDict[str, `Product`]
+            Topologically sorted OrderedDict
+        """
+        deps = []
+        for product_name, product in self.items():
+            deps.extend([(product_name, dependency) for dependency in product.dependencies])
+        topo_sorted_product_lists = tsort.toposort(deps)
+        ordered_product_list = []
+        for group_idx, sort_group in enumerate(topo_sorted_product_lists):
+            self.sorted_groups.append(sort_group)
+            for product_name in sort_group:
+                ordered_product_list.append(self[product_name])
+        self.clear()
+        for product in ordered_product_list:
+            self[product.name] = product
+        self.toposorted = True
+        return self
+
+    def flat_dependencies(self, product: "Product") -> List["Product"]:
+        """Return a flat list of dependencies for this product.
+
+        Parameters
+        ----------
+        product_index : ordered index (from manifest)
+
+
+            Returns:
+                list of `Product`s.
+
+        """
+        resolved: Set[Product] = set()
+        self.flat_dependencies_recurse(resolved, product)
+        return list(resolved)
+
+    def flat_dependencies_recurse(self, resolved: Set["Product"], product: "Product"):
+        """Return a flat list of dependencies for this product.
+
+        Parameters
+        ----------
+        product_index : ordered index (from manifest)
```

Comment:
What is the difference between the "recurse" and non-recurse version? Please make it clear why one would use this or the other.

The doc string describes a parameter product_index that does not exist.

Reply:
The recurse version has been removed